File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/benchmark/BenchmarkStats.kt
package ai.zetic.demo.posemotion.benchmark

import ai.zetic.demo.posemotion.pipeline.BenchmarkSnapshot
import ai.zetic.demo.posemotion.pipeline.FrameTimings
import android.os.SystemClock

/**
 * Rolling per-stage latency + sustained FPS + peak memory.
 * Confined to the inference thread; publish snapshot() to the UI.
 */
class BenchmarkStats {
    private val detect = RollingMean(30)
    private val pose = RollingMean(30)
    private val lift = RollingMean(30)
    private val total = RollingMean(30)
    private val frameStamps = ArrayDeque<Long>()
    private var peakMemoryMB = 0.0

    fun record(t: FrameTimings) {
        detect.push(t.detectMs)
        pose.push(t.poseMs)
        if (t.liftMs > 0) lift.push(t.liftMs)
        total.push(t.totalMs)

        val now = SystemClock.elapsedRealtimeNanos()
        frameStamps.addLast(now)
        while (frameStamps.isNotEmpty() && now - frameStamps.first() > 2_000_000_000L) {
            frameStamps.removeFirst()   // sustained FPS over a 2 s window
        }

        peakMemoryMB = maxOf(peakMemoryMB, MemoryProbe.footprintMB())
    }

    fun resetFps() = frameStamps.clear()

    fun snapshot(): BenchmarkSnapshot {
        var fps = 0.0
        if (frameStamps.size > 1) {
            val spanS = (frameStamps.last() - frameStamps.first()) / 1e9
            if (spanS > 0.2) fps = (frameStamps.size - 1) / spanS
        }
        return BenchmarkSnapshot(
            detectMs = detect.mean,
            poseMs = pose.mean,
            liftMs = lift.mean,
            totalMs = total.mean,
            fps = fps,
            memoryMB = MemoryProbe.footprintMB(),
            peakMemoryMB = peakMemoryMB,
        )
    }
}

private class RollingMean(private val capacity: Int) {
    private val values = ArrayDeque<Double>()
    fun push(v: Double) {
        values.addLast(v)
        while (values.size > capacity) values.removeFirst()
    }
    val mean: Double get() = if (values.isEmpty()) 0.0 else values.sum() / values.size
}

File: apps/Pose-Motion/iOS/PoseMotion/Benchmark/BenchmarkStats.swift
import Foundation
import QuartzCore

struct FrameTimings {
    var detectMs: Double = 0
    var poseMs: Double = 0
    var liftMs: Double = 0
    var totalMs: Double = 0
}

struct BenchmarkSnapshot {
    var detectMs: Double = 0
    var poseMs: Double = 0
    var liftMs: Double = 0
    var totalMs: Double = 0
    var fps: Double = 0
    var memoryMB: Double = 0
    var peakMemoryMB: Double = 0
}

/// Rolling per-stage latency + sustained FPS + peak memory.
/// Confined to the inference queue; publish `snapshot()` to the UI.
final class BenchmarkStats {
    private var detect = RollingMean(capacity: 30)
    private var pose = RollingMean(capacity: 30)
    private var lift = RollingMean(capacity: 30)
    private var total = RollingMean(capacity: 30)
    private var frameStamps: [CFTimeInterval] = []
    private var peakMemoryMB: Double = 0

    func record(_ t: FrameTimings) {
        detect.push(t.detectMs)
        pose.push(t.poseMs)
        if t.liftMs > 0 { lift.push(t.liftMs) }
        total.push(t.totalMs)

        let now = CACurrentMediaTime()
        frameStamps.append(now)
        frameStamps.removeAll { now - $0 > 2.0 }   // sustained FPS over a 2 s window

        peakMemoryMB = max(peakMemoryMB, MemoryProbe.footprintMB())
    }

    func snapshot() -> BenchmarkSnapshot {
        var s = BenchmarkSnapshot()
        s.detectMs = detect.mean
        s.poseMs = pose.mean
        s.liftMs = lift.mean
        s.totalMs = total.mean
        if let first = frameStamps.first, frameStamps.count > 1 {
            let span = CACurrentMediaTime() - first
            if span > 0.2 { s.fps = Double(frameStamps.count - 1) / span }
        }
        s.memoryMB = MemoryProbe.footprintMB()
        s.peakMemoryMB = peakMemoryMB
        return s
    }
}

private struct RollingMean {
    private var values: [Double] = []
    private let capacity: Int
    init(capacity: Int) { self.capacity = capacity }
    mutating func push(_ v: Double) {
        values.append(v)
        if values.count > capacity { values.removeFirst() }
    }
    var mean: Double { values.isEmpty ? 0 : values.reduce(0, +) / Double(values.count) }
}
